### ISSUE TITLE
Cryptic error due to exception during symmetry reduction

### ISSUE THREAD

**Issue by Calvin-L** [tlaplus/tlaplus#506]
I recently saw this output during a model checking run. (Sorry, I am not able to share the input specification, but I might be able to produce a small example on demand now that I understand the issue.)

I think it is fine for TLC to throw an exception in this case. However, the error message is impossible to debug. It provides no information about what went wrong or what I can do to fix it.

```
[...]
Progress(20) at 2020-09-18 15:06:55: 3,269 states generated (3,269 s/min), 1,977 distinct states found (1,977 ds/min), 162 states left on queue.
Error: TLC threw an unexpected exception.
This was probably caused by an error in the spec or model.
See the User Output or TLC Console for clues to what happened.
The exception was a java.lang.RuntimeException
: util.Assert$TLCRuntimeException: Attempted to compare string "abc" with non-string:
TRUE
Error: The error occurred when TLC was evaluating the nested
expressions at the following positions:
    The error call stack is empty.

3956 states generated, 2352 distinct states found, 112 states left on queue.
The depth of the complete state graph search is 23.
Finished in 05s at (2020-09-18 15:06:56)
```

The Java debugger points to this line as the culprit:

https://github.com/tlaplus/tlaplus/blob/e377b760220f41bca21b0cf40f493c162d429d74/tlatools/org.lamport.tlatools/src/tlc2/tool/TLCStateMut.java#L222

The comparison on that line throws an exception because a string and a boolean are being compared. They are compared because
 - My specification contains a function variable whose domain is a set of process IDs and whose range is booleans and strings
 - I have declared process IDs to be a symmetry set
 - Performing symmetry reduction requires comparing permuted versions of the function where the same process ID might map to a boolean or to a string; in this case, it compares
   - `(Alice :> "abc" @@ Bob :> TRUE)` and
   - `(Alice :> TRUE @@ Bob :> "abc")`

It was very unpleasant for me to resort to using the Java debugger to figure out what was happening.

One possible solution: TLC should throw a more helpful message if a comparison fails during symmetry reduction. Perhaps the message could explain that symmetry reduction sometimes requires comparing values associated with different elements of a symmetry set, and encourage users to keep incomparable values separated into different variables.

Another possible solution: perhaps during that stage of symmetry reduction, rather than computing a _single_ minimal state, TLC can compute the smallest non-empty set of minimal states that are mutually incomparable. All of those states would need to be explored separately. Users would not ever have to worry about this kind of exception being introduced by symmetry set reduction.

**Comment by lemmy** [comment]
@Calvin-L For now, improved error reporting seems the way to go.  By the way, do you frequently write specs with a mixture of "types" in the range of a function?

**Comment by Calvin-L** [comment]
Yes, although it is not usually a problem.

More detail:

I often have a per-process variable (i.e. a function from process ID to value) where the value for a particular process can transition from "MISSING" to a meaningful value. Generally I use a model value for the "MISSING" constant, so there is no risk of incomparable values. Sometimes when I am being sloppy, I use a string for the "MISSING" constant---which works fine as long as there is some other variable (like a program counter) that tells me whether the value is missing or not. However, now I know that using a string for the missing value could be risky under symmetry reduction if the meaningful value is a boolean or an integer.

The particular case that led me to this issue is different and might also be interesting. I am working on a fine-grained PlusCal model of a real system to show that it refines a higher-level specification. To help ensure correspondence with the real system, I am structuring the PlusCal code much the same way as the real system, with identically-named procedures. Most of the procedures return a value: either a value read from somewhere or a flag to indicate success or failure. I was using a per-process `rval` variable to store return values from procedures: every procedure writes to `rval[self]` just before `return`. However, some procedures return strings and others return integers and others return records and others return booleans... leading to a very heterogeneous range for `rval` and hence to this issue. For now I have changed the code to have separate `rvalBool` and `rvalStr` and so forth, but I'm interested to hear if you know a better way.

**Comment by lemmy** [comment]
@konnov This is, perhaps, an interesting example for Apalache/type-checking?

**Comment by konnov** [comment]
Yes, it is an interesting case. I hope to finish the new type checker soon, so you could see, whether it makes sense for you.

It is somewhat expected that larger specs require a more structured way of describing data structures. Our type system would reject the code that mixes strings and other kinds of objects. In many languages, you would use algebraic data types to safely mix values of different types. I don't see a clean way of implementing ADTs in the current version of TLA+.

There is an idiom for mixing values of different types in TLA+. For instance, in [Paxos](https://github.com/tlaplus/Examples/blob/master/specifications/Paxos/Paxos.tla) Leslie is mixing messages of different types by introducing records that all have the same field `type`. By checking the value of `type`, the specs figures out the actual record type.

Unfortunately, our type checker will give up there and just compute the most general record type. For instance, the type of the set `{[type |-> "1a", bal : 3], [type |->, acc : 4, bal : 5, val : 6]}` will be inferred as `Set(type: Str, bal: Int, acc: Int, val: Int)`. So the type checker will not help you to keep the clean separation between records of different shapes. If we find a good way of writing something similar to Scala's `match` expressions in TLA+, then we can do more precise type analysis.

**Comment by dgomezferro** [comment]
I ran into this same problem in a different way. I had a PlusCal algorithm where I was initializing process variables to 0, that then had a different type during execution. I did that because some of them were being assigned to Model values so I didn't know what to initialize them to. I also figured it out using a java debugger, a better error message would have been really helpful 😃

**Comment by lemmy** [comment]
> I ran into this same problem in a different way. I had a PlusCal algorithm where I was initializing process variables to 0, that then had a different type during execution. I did that because some of them were being assigned to Model values so I didn't know what to initialize them to. I also figured it out using a java debugger, a better error message would have been really helpful 😃

Did you try the TLA+ debugger?  If yes, why didn't it help diagnose this issue?

**Comment by dgomezferro** [comment]
I wasn't aware of the TLA+ debugger, that's cool! I don't think it would have helped me in this case though, here's what I see when using it:

```
@!@!@STARTMSG 1000:1 @!@!@
TLC threw an unexpected exception.
This was probably caused by an error in the spec or model.
See the User Output or TLC Console for clues to what happened.
The exception was a java.lang.RuntimeException
: Attempted to compare integer 0 with non-integer:
<<>>
@!@!@ENDMSG 1000 @!@!@
@!@!@STARTMSG 2121:1 @!@!@
The behavior up to this point is:
@!@!@ENDMSG 2121 @!@!@
@!@!@STARTMSG 2217:4 @!@!@
1: <Init line 208, col 9 to line 232, col 78 of module draining>
/\ mods_ = (dr0 :> 0 @@ dr1 :> 0)
/\ writeSuccesful = (w0 :> FALSE @@ w1 :> FALSE)
/\ i = (w0 :> 1 @@ w1 :> 1 @@ dr0 :> 1 @@ dr1 :> 1)
/\ mods = ( w0 :> defaultInitValue @@
  w1 :> defaultInitValue @@
  dr0 :> defaultInitValue @@
  dr1 :> defaultInitValue )

[... Some more intermediate states ...] 

@!@!@STARTMSG 2217:4 @!@!@
8: <DRIVER_CATCHUP(dr0) line 315, col 25 to line 325, col 54 of module draining>
/\ mods_ = (dr0 :> 0 @@ dr1 :> 0)
/\ writeSuccesful = (w0 :> FALSE @@ w1 :> FALSE)
/\ i = (w0 :> 1 @@ w1 :> 1 @@ dr0 :> 1 @@ dr1 :> 1)
/\ mods = ( w0 :> defaultInitValue @@
  w1 :> defaultInitValue @@
  dr0 :> defaultInitValue @@
  dr1 :> defaultInitValue )
/\ pc = ( w0 :> "WRITER_START" @@
  w1 :> "WRITER_START" @@
  dr0 :> "DRIVER_COMPUTE_MODS" @@
  dr1 :> "DRIVER_START" )
/\ shardDocs = ( a :> (d0 :> 0 @@ d1 :> 0 @@ d2 :> 0) @@
  b :> (d0 :> 0 @@ d1 :> 0 @@ d2 :> 0) @@
  c :> (d0 :> 0 @@ d1 :> 0 @@ d2 :> 0) )
/\ chunks = {{d0}, {d1, d2}}
/\ destShard = ( w0 :> defaultInitValue @@
  w1 :> defaultInitValue @@
  dr0 :> defaultInitValue @@
  dr1 :> defaultInitValue )
/\ targetShard = (w0 :> 0 @@ w1 :> 0)
/\ chunkToMove = (dr0 :> {d0} @@ dr1 :> 0)
/\ ownership = ({d0} :> a @@ {d1, d2} :> a)
/\ destShard_ = (dr0 :> b @@ dr1 :> 0)
/\ sourceShard = (dr0 :> a @@ dr1 :> 0)
/\ docsToCopy = (dr0 :> {} @@ dr1 :> {})
/\ targetChunk = (w0 :> 0 @@ w1 :> 0)
/\ writesBlocked = ({d0} :> FALSE @@ {d1, d2} :> FALSE)
/\ allDocs = (d0 :> 0 @@ d1 :> 0 @@ d2 :> 0)
/\ moveInProgress = ({d0} :> TRUE @@ {d1, d2} :> FALSE)
/\ pendingWrites = ({d0} :> <<>> @@ {d1, d2} :> <<>>)
/\ write = (w0 :> 0 @@ w1 :> 0)
/\ stack = (w0 :> <<>> @@ w1 :> <<>> @@ dr0 :> <<>> @@ dr1 :> <<>>)

@!@!@ENDMSG 2217 @!@!@
@!@!@STARTMSG 2103:1 @!@!@
The error occurred when TLC was evaluating the nested
expressions at the following positions:
    The error call stack is empty.

@!@!@ENDMSG 2103 @!@!@
```
